[Clang][SYCL] Introduce clang-sycl-linker to link SYCL offloading device code (Part 1 of many)

asudarsa opened the pull request:
This PR is one of the many PRs in the SYCL upstreaming effort focusing on device code linking during the SYCL offload compilation process. RFC: https://discourse.llvm.org/t/rfc-offloading-design-for-sycl-offload-kind-and-spir-targets/74088

In this PR, we introduce a new tool that will be used to perform device code linking for SYCL offload kind. It accepts SYCL device objects in LLVM IR bitcode format and will generate a fully linked device object that can then be wrapped and linked into the host object.

A primary use case for this tool is to perform device code linking for objects with SYCL offload kind inside the clang-linker-wrapper. It can also be invoked via clang driver as follows:

`clang --target=spirv64 --sycl-link input.bc`

Device code linking for SYCL offloading kind has a number of known quirks that makes it difficult to use in a unified offloading setting. Two of the primary issues are:
1. Several finalization steps are required to be run on the fully-linked LLVM IR bitcode to gaurantee conformance to SYCL standards. This step is unique to SYCL offloading compilation flow.
2. SPIR-V LLVM Translator tool is an extenal tool and hence SPIR-V IR code generation cannot be done as part of LTO. This limitation will be lifted once SPIR-V backend is available as a viable LLVM backend.

Hence, we introduce this new tool to provide a clean wrapper to perform SYCL device linking.

Co-Author: Michael Toguchi
Thanks

asudarsa committed:
[Clang][SYCL] Introduce clang-sycl-link-wrapper to link SYCL offloading device code

This PR is one of the many PRs in the SYCL upstreaming effort focusing on
device code linking during the SYCL offload compilation process.
RFC: https://discourse.llvm.org/t/rfc-offloading-design-for-sycl-offload-kind-and-spir-targets/74088

In this PR, we introduce a new tool that will be used to perform device code
linking for SYCL offload kind. It accepts SYCL device objects in LLVM IR bitcode
format and will generate a fully linked device object that can then be wrapped
and linked into the host object.

A primary use case for this tool is to perform device code linking for objects
with SYCL offload kind inside the clang-linker-wrapper.
It can also be invoked via clang driver as follows:

`clang --target=spirv64 --sycl-link input.bc`

Device code linking for SYCL offloading kind has a number of known quirks that
makes it difficult to use in a unified offloading setting.
Two of the primary issues are:
1. Several finalization steps are required to be run on the fully-linked LLVM
IR bitcode to gaurantee conformance to SYCL standards. This step is unique to
SYCL offloading compilation flow.
2. SPIR-V LLVM Translator tool is an extenal tool and hence SPIR-V IR code
generation cannot be done as part of LTO. This limitation will be lifted once
SPIR-V backend is available as a viable LLVM backend.

Hence, we introduce this new tool to provide a clean wrapper to perform SYCL
device linking.

Thanks

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
This PR is the first in the list of multiple PRs that will be submitted to add SYCL offloading support using the new offloading model. Upcoming PRs include:
1. SYCL finalization steps that will run after llvm-link will be added to the linking clow inside clang-sycl-link-wrapper. One of the finalization steps is device code splitting.
2. Changes to clang-linker-wrapper to invoke the clang-sycl-link-wrapper (via the clang driver call) for SYCL offloading case.
3. AOT compilation support for Intel, AMD and NVidia GPUs

Thanks
Sincerely

asudarsa commented:
@jhuber6

Can you please take a look at this PR when convenient? Thanks for all your inputs in getting this PR ready.

Sincerely

asudarsa commented on clang/tools/clang-sycl-link-wrapper/ClangSYCLLinkWrapper.cpp:
```diff
@@ -0,0 +1,530 @@
+
+  TranslatorArgs.push_back(Args.MakeArgString(UnknownIntrinsics));
+
+  // Disable all the extensions by default
```
List of SPIR-V extensions is 'fluid' and will be updated in an upcoming commit.

Thanks

asudarsa committed:
Update code comment

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

jhuber6 commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6728,7 +6728,10 @@ def fsycl : Flag<["-"], "fsycl">,
   Visibility<[ClangOption, CLOption]>,
   Group<sycl_Group>, HelpText<"Disables SYCL kernels compilation for device">;
-
+def sycl_link : Flag<["--"], "sycl-link">, Flags<[HelpHidden]>,
```
Why is this necessary? Shouldn't it do this automatically if we pass in `.o` files or use `-Xlinker foo.c`?

asudarsa replied:
We use this option to select between spirv-linker which is the standard linking tool for spir-v IR (but not mature enough for us to use it in our pipeline) and the clang-sycl-linker.

Thanks

jhuber6 commented on clang/lib/Driver/ToolChains/SPIRV.cpp:
```diff
@@ -95,7 +95,19 @@ void SPIRV::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+SPIRVToolChain::SPIRVToolChain(const Driver &D, const llvm::Triple &Triple,
+                               const ArgList &Args)
+    : ToolChain(D, Triple, Args) {
+  NativeLLVMSupport = Args.hasArg(options::OPT_sycl_link);
```
Okay I'm guessing this is to separate this from an existing SPIR-V link step? We could possibly put that in a different toolchain.

asudarsa replied:
We can eventually move this to SYCL toolchain when it is available upstream.

Thanks

jhuber6 replied:
Alright, we can put that as a `TODO` then.

jhuber6 commented on clang/lib/Driver/ToolChains/SPIRV.cpp:
```diff
@@ -95,7 +95,19 @@ void SPIRV::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+  // Use of --sycl-link will call the clang-sycl-link-wrapper instead of
+  // the default linker (spirv-link).
+  if (Args.hasArg(options::OPT_sycl_link))
+    Linker = ToolChain.GetProgramPath("clang-sycl-link-wrapper");
```
I'm not sure I'd call this a wrapper since it doesn't really wrap around anything as far as I'm aware, it's more of a utility that creates a SYCL image from some LLVM-IR. Maybe just `linker`?

asudarsa replied:
That sounds better. Will add the change. Thanks

asudarsa replied:
Made the change in 679436acf28a4da805dfbdf388ca1c9ce692ad4d

Thanks

jhuber6 commented on clang/test/Driver/Inputs/libsycl-complex.bc:
Why do we need these? We can create empty bitcode files inside the tests using `clang`. (Also I'm wondering if there's an existing empty bitcode file).

asudarsa replied:
Good point. I will create them inside the test case. I was trying to 'mimic' device library file usage. I suppose it's overkill. Thanks.

jhuber6 replied:
Testing like that is fine, it's just that you can do `clang_cc1 -emit-llvm %s -o %t.bc` or something.

asudarsa replied:
Modified the test to just use empty files. Change can be found in 307c74ea4f910a7698a6084db710ff7b699c5c93

Thanks

jhuber6 commented on clang/tools/clang-sycl-link-wrapper/ClangSYCLLinkWrapper.cpp:
```diff
@@ -0,0 +1,530 @@
+#include "SYCLLinkOpts.inc"
+#undef PREFIX
+
+static constexpr OptTable::Info InfoTable[] = {
```
The thing that's special about this case is that it forwards arguments it doesn't understand via a flag. We could possibly make a utility for that.

jhuber6 commented on clang/tools/clang-sycl-link-wrapper/ClangSYCLLinkWrapper.cpp:
```diff
@@ -0,0 +1,528 @@
+  return *OutFileOrErr;
+}
+
+const SmallVector<std::string> SYCLDeviceLibNames = {
```
If these are part of the ToolChain they should be passed in by default when the `clang-sycl-linker` is created. Pretty similar to the CUDA libdevice and ROCm Device Libs except we're doing it at the link step instead of hacking it in per TU with `-mlink-builtin-bitcode`. Same reason that `clang` is responsible for putting `-lc` in the link job and not `ld.lld`.

asudarsa replied:
Added a flag that the driver can use to pass in the libdevice libs in 307c74ea4f910a7698a6084db710ff7b699c5c93

However, we have not upstreamed the device library sources and build process. Will pass the files via the driver once available.

Thanks

jhuber6 commented on clang/tools/clang-sycl-link-wrapper/ClangSYCLLinkWrapper.cpp:
```diff
@@ -0,0 +1,528 @@
+/// 'Args' encompasses all arguments required for linking and wrapping device
+/// code and will be parsed to generate options required to be passed into the
+/// llvm-spirv tool.
+static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef File,
```
If any of these are external, we should also pass a path in from clang like how we handle CUDA / ROCm. If these are internal tools then don't worry about it since we just search next to `clang`.

asudarsa replied:
Added a new option to pass the llvm-spirv path to the clang-sycl-linker tool.

Thanks

jhuber6 reviewed (commented):
First pass, I like the approach overall.

asudarsa commented:
Hi @jhuber6 

Thanks much for providing your feedback. I will address all your comments in an upcoming commit.

Sincerely

asudarsa committed:
Renamed the tool to be called clang-sycl-linker

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Address review comments

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Remove unused files

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Earlier test fail due to lack of new tool in clang/test/lit.cfg.py.
Added it now. 

Thanks

asudarsa commented:
Hi @jhuber6 

I have addressed all the concerns to my best ability. Thanks very much for your guidance and feedback thus far.

Thanks
Sincerely

jhuber6 commented on clang/docs/ClangSYCLLinker.rst:
```diff
@@ -0,0 +1,82 @@
+    -help-hidden                  Display all available options
+    -help                         Display available options (--help-hidden for more)
+    --library-path=<dir>          Set the library path for SYCL device libraries
+    --device-libs=<value>         A comma separated list of device libraries that are linked during the device link
```
Why are device libs and regular input distinct?

asudarsa replied:
Regular inputs for SYCL offloading are linked using default settings whereas SYCL device libs are linked with the '--only-needed' flag.

Thanks

jhuber6 replied:
That's kind of the behavior of static libraries, and LTO could just optimize them out if they're unused. But I suppose it's not terribly important to get it perfect on the first pass.

asudarsa committed:
Fix documenetation error. Add TODO for --sycl-link usage. Add couple of changes missed out from earlier commit.

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

jhuber6 reviewed (approved):
Seems reasonable, more work can be done in follow ups. Please wait until the CI is green before merging anything.

asudarsa committed:
Fix minor bug

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

llvm-beanz reviewed (commented):
Why is this a clang-based tool? Other than using clang/Basic to get the clang version (which could also be retrieved from LLVM), this doesn't seem to have any dependencies on Clang.

As I'm seeing the SYCL patches coming in I'm getting concerned that the architecture of the SYCL compiler doesn't align with LLVM and Clang's design.

llvm-beanz reviewed (dismissed):
I'd really like to understand why this is a clang-level tool and get feedback from @AaronBallman before this goes further.

jhuber6 commented:
> Why is this a clang-based tool? Other than using clang/Basic to get the clang version (which could also be retrieved from LLVM), this doesn't seem to have any dependencies on Clang.
> 
> As I'm seeing the SYCL patches coming in I'm getting concerned that the architecture of the SYCL compiler doesn't align with LLVM and Clang's design.

Realistically it doesn't need to be, despite the fact that clang wants to use it. The goal is to just have a tool that clang can use as a normal toolchain linker so we can reuse as much as possible when it comes to creating link jobs. We could stash this in `llvm/` and it wouldn't make much of a difference. Also we can probably just use the LLVM Linker API instead of `llvm-link`. Forgot to mention that as a to-do.

jhuber6 commented:
> > Why is this a clang-based tool? Other than using clang/Basic to get the clang version (which could also be retrieved from LLVM), this doesn't seem to have any dependencies on Clang.
> > As I'm seeing the SYCL patches coming in I'm getting concerned that the architecture of the SYCL compiler doesn't align with LLVM and Clang's design.
> 
> Realistically it doesn't need to be, despite the fact that clang wants to use it. The goal is to just have a tool that clang can use as a normal toolchain linker so we can reuse as much as possible when it comes to creating link jobs. We could stash this in `llvm/` and it wouldn't make much of a difference. Also we can probably just use the LLVM Linker API instead of `llvm-link`. Forgot to mention that as a to-do.

I don't know if there's some existing heuristic for what clang tools do, but generally I find it to either be something that the user interfaces in, or something that clang uses when it's building job actions. This is the latter.

llvm-beanz commented:
> The goal is to just have a tool that clang can use as a normal toolchain linker so we can reuse as much as possible when it comes to creating link jobs.

Clang can use `lld`, `llvm-ar`, and `dsymutil` (among others), so being part of Clang certainly isn't necessary for the clang driver to invoke it.

>We could stash this in `llvm/` and it wouldn't make much of a difference. Also we can probably just use the LLVM Linker API instead of `llvm-link`. Forgot to mention that as a to-do.

I'd be real interested to know if this should be part of `llvm-link` or `lld`. As a general rule we try not to accept streams of patches with big `TODO` markings to clean it all up after it lands.

jhuber6 commented:
> Clang can use `lld`, `llvm-ar`, and `dsymutil` (among others), so being part of Clang certainly isn't necessary for the clang driver to invoke it.

Yes, but generally we've just stashed stuff in `clang` if we didn't plan to expose them as general utilities, since it's not like this tool is of any worth to people without `clang` enabled in their projects list.



> I'd be real interested to know if this should be part of `llvm-link` or `lld`. As a general rule we try not to accept streams of patches with big `TODO` markings to clean it all up after it lands.

I doubt it, SYCL's target is SPIR-V which isn't a traditional linked output. We could *possibly* use LTO and (once the SPIR-V backend is complete) use `-Wl,--lto-emit-llvm` to just get it to spit out the result of that and give up, but that's a little hacky. This is only using `llvm-link` to link bitcode files, which can just be done through the `Linker` API, nothing that should be inside of the linker API itself. As for TODO's, the idea is that you hopefully get the basic functionality so people can use it and then start adding more features, instead of having one gigantic patch that no one wants to review.

llvm-beanz commented:
> I don't know if there's some existing heuristic for what clang tools do, but generally I find it to either be something that the user interfaces in, or something that clang uses when it's building job actions. This is the latter.

Generally clang-based tools have dependencies on Clang. This tool does not (other than pulling the version number which could come from LLVM).

It only pulls in the version number. Having the tool higher in the dependency graph avoids unnecessary dependencies, improves build/test/rebuild cycles and is generally in line with how things are done.

jhuber6 commented:
> Generally clang-based tools have dependencies on Clang. This tool does not (other than pulling the version number which could come from LLVM).

We can move this to `llvm` if it's a blocking issue, doesn't really make much of a difference (I forget if this patch adds it to the list of test dependencies, should do that as well)

llvm-beanz commented:
I'm also concerned that there are no tests for the new tool. There are tests for the clang driver changes, but none for the new tool. The new tool seems to be just a wrapper around llvm-link, which does make me think this should probably just be folded into llvm-link. Understanding the direction here would be helpful.

jhuber6 commented:
> I'm also concerned that there are no tests for the new tool. There are tests for the clang driver changes, but none for the new tool. The new tool seems to be just a wrapper around llvm-link, which does make me think this should probably just be folded into llvm-link. Understanding the direction here would be helpful.

There is one, but it should probably be expanded [clang-sycl-linker-test.cpp](https://github.com/llvm/llvm-project/pull/112245/files#diff-373ef22afa527adac2fdce560a26553ed7280e4bc0b75f01ff77e1d64140f491). The purpose is to behave like a Toolchain linker so that when we do `--target=spirv64-whatever` it will create a valid block for that target as is customary with `clang` cross-compiling.

I don't know exactly what SYCL needs that's different from `llvm-link`, maybe the others can chime in because I thought that it required the backend being run on it. I think the reason this doesn't use the LTO interface right now is because the SPIR-V target isn't done? I'm sure some of the SYCL team can explain what they're doing but I assume the goal is

foo.bc bar.bc -> image.spv

llvm-beanz commented:
It seems to me like this should be done through the LTO interface and driven through `lld`. I can understand if an intermediate step is required while the SPIR-V backend is under development, but clang shouldn't be in the business of linking, and generally neither should llvm-link.

I'd be curious for @MaskRay's perspective too.

jhuber6 commented:
> It seems to me like this should be done through the LTO interface and driven through `lld`. I can understand if an intermediate step is required while the SPIR-V backend is under development, but clang shouldn't be in the business of linking, and generally neither should llvm-link.
> 
> I'd be curious for @MaskRay's perspective too.

I don't think `lld` makes sense, but you can definitely use the LTO interface inside of this tool to create a similar effect. That should also get rid of the need for this `--device-library` thing because then it'll just internalize anything that's unused. That is, if the SYCL folks write their runtimes correctly with hidden visibility on device functions.

asudarsa committed:
Add clang-sycl-linker to CLANG_TEST_DEPS list

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

llvm-beanz commented:
> I don't think `lld` makes sense, but you can definitely use the LTO interface inside of this tool to create a similar effect. 

Why not? `lld` is a toolchain linker (which you said this is supposed to emulate), and it is an interface to the LTO interface... so it seems to me like `lld` is the right place to do this.

jhuber6 commented:
> > I don't think `lld` makes sense, but you can definitely use the LTO interface inside of this tool to create a similar effect.
> 
> Why not? `lld` is a toolchain linker (which you said this is supposed to emulate), and it is an interface to the LTO interface... so it seems to me like `lld` is the right place to do this.

I guess the closest analogue would be the WASM target?  It seems like overkill since the `ld.lld` source code doesn't really share much code and  this target pretty much only consumes LLVM-IR. I guess @MaskRay can vote on what makes sense here, since AFAIK (I'm not a SYCL or SPIR-V guy) this is mostly just wrapping some toolchain specific magic around LTO linking. You could totally do it even with the ELF target if you just pass `-Wl,--lto-emit-llvm`, I think the SPIR-V target might work soonish so maybe that's a reasonable option.

llvm-beanz commented:
> I guess the closest analogue would be the WASM target?

LLD has a wasm target.

> It seems like overkill since the `ld.lld` source code doesn't really share much code and this target pretty much only consumes LLVM-IR.

I think LLD also needs to eventually gain a SPIR-V target to support linking SPIR-V binaries, because SPIR-V does support linkage at the SPIR-V level not just the LTO IR level (see: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_linkage). I certainly know that my team will want to use lld as the LTO driver for DXIL and SPIR-V when targeting DirectX and Vulkan with HLSL.

IIRC we've also previously discussed having all the offload languages use LTO pipelines by default, it seems to make sense that we would also use `lld` as the LTO driver.

jhuber6 commented:
> LLD has a wasm target.

Exactly what I meant.

> I think LLD also needs to eventually gain a SPIR-V target to support linking SPIR-V binaries, because SPIR-V does support linkage at the SPIR-V level not just the LTO IR level (see: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_linkage). I certainly know that my team will want to use lld as the LTO driver for DXIL and SPIR-V when targeting DirectX and Vulkan with HLSL.

Yeah I would like if you could link direct SPIR-V, i think right now that's done with an external tool? Using `lld` would be the logical place to put that. I just mean that we can introduce simpler solutions and then work to more complete ones. That seems like a more complete issue than what we're doing here though. (Also I don't have the most complete picture for what special things SYCL needs to do here. Hopefully they can chime in but IIRC it has to do with merging several images / binaries into some logical set so it's not as simple as linking?).

llvm-beanz commented:
I think we could all benefit from some documentation describing how the SYCL compiler flow is intended to work, what tools are added/modified, and what the expected outputs are at each compiler phase. Without some idea of the architecture of what is being built changes like this and #110771 don't make sense and are really difficult to review.

This isn't like introducing a new backend or something where we have a defined and understood architecture to evaluate against. We're getting patches that only represent small parts of the whole without any context to understand how it will all come together in the end.

What worries me is that from the parts I see, it doesn't align with any of the community design principles.

jhuber6 commented:
> I think we could all benefit from some documentation describing how the SYCL compiler flow is intended to work, what tools are added/modified, and what the expected outputs are at each compiler phase. Without some idea of the architecture of what is being built changes like this and #110771 don't make sense and are really difficult to review.
> 
> This isn't like introducing a new backend or something where we have a defined and understood architecture to evaluate against. We're getting patches that only represent small parts of the whole without any context to understand how it will all come together in the end.
> 
> What worries me is that from the parts I see, it doesn't align with any of the community design principles.

From what I understand, they want to integrate SYCL linking into the existing linker wrapper framework. That tool basically extracts embedded offload binaries and then calls `clang --target=${target} extracted.o -o image` and then embeds that into the runtime. They wanted to know how to do that so my suggestion was to make something that looked like a linker so that control flow worked. I'm not familiar with the other serialization format however.

asudarsa commented:
Hi @llvm-beanz and @jhuber6 

Thanks so much for all the feedback and discussion here. I am very much grateful for both your time and insights here. I will try to give my answers in this comment.
An overview of the SYCL compilation flow (with a bit more emphasis on the device code linking phases) can be seen in the RFC that we submitted earlier. 
RFC: https://discourse.llvm.org/t/rfc-offloading-design-for-sycl-offload-kind-and-spir-targets/74088

As shown in the RFC, SYCL device code linking process involves the following steps:
1. Linking of all LLVM IR bitcode device objects found in user code. llvm-link is used here.
2. Linking of fully linked LLVM IR device bitcode found in user code (output of the previous step) with SYCL specific device library files. llvm-link is used here.
3. Post-link steps: Some steps are required to conform to SYCL specification. Some kernels have kernel metadata (known as SYCL properties) that need to be communicated to SYCL runtime. This requires processing of LLVM IR metadata and generating some information that can be embedded into the device image and then processed by SYCL runtime. Also, we support device code splitting which is mostly done to improve run-time performance and also ensure SYCL spec conformance in some cases.
4. For SPIR-V based JIT and AOT targets, SPIR-V IR code generation is then performed using an external tool (llvm-spirv)
5. For AOT targets (both SPIR-V and non SPIR-V), we then invoke backend compilation tools (For e.g. ocloc for Intel graphics).

Quite a few of these steps have been fashioned to work around the fact that we do not have a regular SPIR-V backend in LLVM repo. Once the SPIR-V backend is ready to be included in compilation flows, we can remove the dependence on llvm-link and external tools like llvm-spirv and use the backend for LTO based linking and SPIR-V code generation.

We initially envisioned this flow to be embedded into clang-linker-wrapper. Due to specific variations from the community flow, especially the need for SYCL-specific post-link steps, we decided to move the logic into a stand-alone tool to do SYCL specific linking.

A primary reason for making this a clang-based tool was that it is quite similar in its functionality to existing tools like clang-linker-wrapper and clang-nvlink-wrapper. As @jhuber6 mentioned, I think it can be easily moved to llvm/tools if that's a blocking issue.

As far as testing support goes, the test that has been added in this PR 'clang-sycl-linker-test.cpp' covers the testing of an very initial flow (linking of source llvm IR device bitcodes, linking of device library files, and spir-v code generation). This test is expected to grow as we add other pieces (post-link steps, AOT compilation support, etc).

I can provide pointers to some extended documentation if the RFC is not in sufficient detail.

Thanks again for great review comments and feedback here.
Sincerely

llvm-beanz reviewed (commented):
At a minimum I think this change needs more tests. There's a lot of code added with very minimal test coverage.

I'm uncomfortable with adding a design that is effectively a workaround with an unwritten "TODO" to fix it up later, but I also don't want to needlessly block things.

From the discourse post and everything I've found reading about the SYCL tooling, it seems to me like this should really just all be integrated into LLD and performed with the linking phase. It seems like a huge waste of IO to read objects, rip out device-specific bits, process that separately, then read the objects again in another process to link the host bits, then in a third process read the linked host and device bits and combine them...

jhuber6 commented:
> From the discourse post and everything I've found reading about the SYCL tooling, it seems to me like this should really just all be integrated into LLD and performed with the linking phase. It seems like a huge waste of IO to read objects, rip out device-specific bits, process that separately, then read the objects again in another process to link the host bits, then in a third process read the linked host and device bits and combine them...

Doing all of this in `lld` is something I floated around at the inception of the offload handling but decided it was more trouble than it was worth. You can think of this flow as a sort of linker plugin, where we preprocess a bunch of input files and then give a result back. In this case the result is an object with the embedded device code and the runtime calls necessary to register it. I didn't go this route for two reasons. First, it forces the user to change their host linker which people don't like. Second, we would need to go through great lengths to make the NVPTX target work because they don't use `ld.lld` and their toolchain is proprietary. i doubted it was going to be a very popular option to make `ld.lld` a frontend for `nvlink`, but we could go that route if so inclined, it's the same thing as another tool I wrote to work around NVIDIA's linker being subpar.

asudarsa commented:
> At a minimum I think this change needs more tests. There's a lot of code added with very minimal test coverage.
> 
> I'm uncomfortable with adding a design that is effectively a workaround with an unwritten "TODO" to fix it up later, but I also don't want to needlessly block things.
> 
> From the discourse post and everything I've found reading about the SYCL tooling, it seems to me like this should really just all be integrated into LLD and performed with the linking phase. It seems like a huge waste of IO to read objects, rip out device-specific bits, process that separately, then read the objects again in another process to link the host bits, then in a third process read the linked host and device bits and combine them...

Hi @llvm-beanz 

I will revisit the testing coverage aspect and try to add more tests.

Thanks

asudarsa committed:
Add more testing

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Remove test

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
clang/test/Driver is not the correct place to be testing for tool availability.

Thanks

asudarsa commented:
Hi @llvm-beanz 

Thanks so much for providing your valuable time and feedback thus far. I have addressed the concerns raised to the best of my ability. I have added more testing inside 'clang/test/Driver/clang-sycl-linker-test.cpp'. It will be great if you can take another look when convenient and let me know if there are additional concerns.

Thanks
Sincerely

P.S: About the location of this tool under clang/tools, I added the tool here as it is somewhat similar in functionality to tools like clang-linker-wrapper and clang-nvlink-wrapper. If there is strong opposition, we can surely move it to llvm/tools.

asudarsa commented:
@bader, thanks for adding the SYCL label. Can you please provide your feedback for this PR when convenient? For some reason, I am not able to add reviewers.

Thanks

jhuber6 commented on clang/tools/clang-sycl-linker/SYCLLinkOpts.td:
```diff
@@ -0,0 +1,55 @@
+def arch : Separate<["--", "-"], "arch">,
+  HelpText<"Specify the name of the target architecture.">;
+
+def g : Flag<["-"], "g">, HelpText<"Specify that this was a debug compile.">;
+def debug : Flag<["--"], "debug">, Alias<g>;
```
Probably not necessary here, probably copied from the NVPTX case where we needed to handle it separately because `-O2` is an error if the PTX was created with `-g`.

asudarsa replied:
We 'might' need this at some stage. But I will remove this for now. Thanks

jhuber6 commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -0,0 +1,42 @@
+// DEVLIBS-NEXT: "{{.*}}llvm-spirv{{.*}}" {{.*}}-o a.spv [[SECONDLLVMLINKOUT]].bc
+//
+// Test a simple case with .o (fat object) as input.
+// TODO: Remove this test once fat object support is added.
```
Yes, that's correct. Currently trying to make the linker wrapper a bit leaner, but the expected flow should be
```
fatbin.o -> dev1.o, dev2.o
  - clang --target=TRIPLE -o image dev1.o dev2.o
```
My general assertion is that every target should have a way to get a usable image via a `clang --target` invocation, which should be equipped to handle normal linker stuff.

asudarsa replied:
Currently, the flow is this. clang-inker-wrapper reads the fat object and extracts bitcode files. These bitcode files are then fed to clang++ as inputs.
We may eventually want to pass the fat object as is and let the clang-sycl-linker extract the bitcode file.

Thanks

bader replied:
> We may eventually want to pass the fat object as is and let the clang-sycl-linker extract the bitcode file.

We don't want that. We want to link object files with the code for a single target (i.e. offload device target). We don't want `clang-sycl-linker` tool to handle **fat** objects - only **regular** objects. 

I suggest we update the comment by replacing "fat object" to just "object".

jhuber6 replied:
Yes, the idea is that the device toolchain should work in the same way that `clang --target=aarch64` gives you a functioning Arm executable.

asudarsa replied:
Adjusted comment. Thanks

jhuber6 commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -0,0 +1,500 @@
+  if (CmdArgs.empty())
+    return;
+
+  llvm::errs() << " \"" << CmdArgs.front() << "\" ";
+  llvm::errs() << llvm::join(std::next(CmdArgs.begin()), CmdArgs.end(), " ")
+               << "\n";
```
I considered this debugging and AFAIK LLVM uses `errs()` for debugging messages.

asudarsa commented on clang/tools/clang-sycl-linker/SYCLLinkOpts.td:
```diff
@@ -0,0 +1,55 @@
+def dry_run : Flag<["--", "-"], "dry-run">, Flags<[WrapperOnlyOption]>,
+  HelpText<"Print generated commands without running.">;
+
+def spirv_dump_device_code_EQ : Joined<["--", "-"], "spirv-dump-device-code=">,
+  Flags<[WrapperOnlyOption]>,
+  HelpText<"Path to the folder where the tool dumps SPIR-V device code. Other formats aren't dumped.">;
```
We do intend to support a user-visible flag at clang-level to allow dumping of SPIR-V device code to a user-specific directory. This option will be fed by that.

Thanks

bader commented on clang/test/Driver/sycl-link-spirv-target.cpp:
```diff
@@ -0,0 +1,13 @@
+// LINK: "{{.*}}clang-sycl-linker{{.*}}" "{{.*}}.bc" "-o" "a.out"
+//
+// Test that -Xlinker options are being passed to clang-sycl-linker.
+// RUN: %clangxx -### --target=spirv64 --sycl-link -Xlinker --llvm-spirv-path=/tmp \
+// RUN:   -Xlinker --library-path=/tmp -Xlinker --device-libs=lib1.bc,lib2.bc %t.bc 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=XLINKEROPTS
+// XLINKEROPTS: "{{.*}}clang-sycl-linker{{.*}}" "--llvm-spirv-path=/tmp" "--library-path=/tmp" "--device-libs=lib1.bc,lib2.bc" "{{.*}}.bc" "-o" "a.out"
```
I mean the code in this file - lines 5-7.

asudarsa replied:
Ah. Sorry. Misinterpretation. I got it. 
I agree. I was incrementally adding these tests. I can drop the first one. 

Thanks

asudarsa commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -0,0 +1,500 @@
+#undef OPTION
+};
+
+class WrapperOptTable : public opt::GenericOptTable {
```
Will change this. Thanks. Also. I will change WrapperOnlyOption to LinkerOnlyOption.

Thanks

asudarsa committed:
Some non-functional refactoring and extra check to not call llvm-link when only one input is present

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @bader 

Thanks very much for your feedback. I have addressed the concerns in the latest commit.
Thanks to @jhuber6 for some responses as well.

Sincerely

bader reviewed (approved):
LGTM except a couple of minor nits.

asudarsa committed:
Update test to specify correct inputs to clang-sycl-linker and update some comments

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

bader commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -0,0 +1,48 @@
+// DEVLIBS-NEXT: "{{.*}}llvm-link{{.*}}" -only-needed [[FIRSTLLVMLINKOUT]].bc {{.*}}lib1.bc {{.*}}lib2.bc -o [[SECONDLLVMLINKOUT:.*]].bc --suppress-warnings
+// DEVLIBS-NEXT: "{{.*}}llvm-spirv{{.*}}" {{.*}}-o a.spv [[SECONDLLVMLINKOUT]].bc
+//
+// Test a simple case with .o (fat object) as input.
+// TODO: Remove this test once fat object support is added.
```
```suggestion
// Test a simple case with .o (object) as input.
// TODO: Remove this test once object support is added.
```

asudarsa replied:
Can I please update this in one of the upcoming PRs? I do not want to trigger another round of testing here.

Thanks

jhuber6 commented:
> I'd prefer we have SYCL linking run through LLD rather than the linker-wrapper tool approach, but I don't feel strongly enough to block this.
I see this as a temporary hack around not having a proper SPIR-V linker / backend in LLVM yet. It's less of a linker-wrapper tool and more of my assertions that every single target should be able to be targeted directly via `clang --target=...` which then lets the linker wrapper just do the part where it extracts GPU code from fat binaries, invokes `clang` on each to get an executable, then wraps that in the runtime's registration code.

asudarsa commented:
All the failures seem to be related to missing dependencies. I am working on a fix. Thanks

jhuber6 commented:
Should be fixed in https://github.com/llvm/llvm-project/commit/0d499f9043fed14ff8c7f9e24e19206c93aee5dd

asudarsa commented:
Sanitizer test failures should be resolved by https://github.com/llvm/llvm-project/pull/114488 (under review).
Thanks to @sarnex for the timely fix.

Sincerely

kstoimenov commented:
@asudarsa looks like this is causing memory leaks in Sanitizer bots:
https://lab.llvm.org/buildbot/#/builders/169/builds/4835/steps/13/logs/stdio
